File: CheckListSQL/CheckListSQL/ItemDetailViewController.m
//
//  ItemDetailViewController.m
//  Checklists_Code
//
//  Created by wangx on 15/6/24.
//  Copyright (c) 2015年 wxiang1991. All rights reserved.
//

#import "ItemDetailViewController.h"
#import "ChecklistItem.h"
#import "ChecklistItemService.h"

@interface ItemDetailViewController()

@property (nonatomic, strong) ChecklistItemService *checklistItemService;

@end

@implementation ItemDetailViewController{
    NSString *_context;
    BOOL _shouldRemind;
    BOOL _datePickerVisible;
    NSDate *_dueDate;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    UIBarButtonItem *cancelBarButton=[[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(cancel)];
    self.navigationItem.leftBarButtonItem=cancelBarButton;
    
    self.doneBarButton=[[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(done)];
    self.navigationItem.rightBarButtonItem=self.doneBarButton;
    
    self.checklistItemService=[[ChecklistItemService alloc]init];
    
    if (self.itemToEdit.context != nil) {
        self.title=@"Edit Item";
        _context=self.itemToEdit.context;
        _shouldRemind=self.itemToEdit.shouldRemind;
        _dueDate=self.itemToEdit.dueDate;
    }else{
        self.title=@"Add Item";
        self.switchControl.on=NO;
        _dueDate=[NSDate date];
    }
}


-(void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:YES];
    //称为第一响应者
    [self.textField becomeFirstResponder];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}
//返回每一个section中的row数
-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    if(section ==0 && _datePickerVisible){
        return 4;
    }else{
        return 3;
    }
}
//新建UITableViewCell
-(UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    static NSString *CellIdentifer=@"Cell";
    UITableViewCell *cell=[tableView dequeueReusableCellWithIdentifier:CellIdentifer];
    if (cell==nil) {
        cell=[[UITableViewCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifer];
    }
    //填写待办事项的名字
    if(indexPath.row==0){
        self.textField=[[UITextField alloc]initWithFrame:CGRectMake(10, 7, 300, 30)];
        self.textField.placeholder=@"Checklist of Name";
        self.textField.text=_context;
        [cell.contentView addSubview:self.textField];
    }else if (indexPath.row==1){//设置是否进行提醒
        UILabel *label=[[UILabel alloc]init];
        label.frame=CGRectMake(10, 7, 100, 30);
        label.text=@"Remind Me";
        label.textAlignment=NSTextAlignmentCenter;
        [cell.contentView addSubview:label];
        
        self.switchControl=[[UISwitch alloc]init];
        self.switchControl.frame=CGRectMake(250, 6, 51, 31);
        self.switchControl.on=_shouldRemind;
        [cell.contentView addSubview:self.switchControl];
        
    }else if(indexPath.row==2){//设置是否记录日期
        UILabel *label=[[UILabel alloc]init];
        label.frame=CGRectMake(10, 7, 80, 30);
        label.text=@"Due Date";
        label.textAlignment=NSTextAlignmentCenter;
        [cell.contentView addSubview:label];
        
        self.dueDateLabel=[[UILabel alloc]init];
        self.dueDateLabel.frame=CGRectMake(100, 7, 220, 30);
        self.dueDateLabel.textAlignment=NSTextAlignmentCenter;
        [cell.contentView addSubview:self.dueDateLabel];
        [self updateDueDateLabel];
    }else{
        cell.selectionStyle=UITableViewCellSelectionStyleNone;
        UIDatePicker *datePicker=[[UIDatePicker alloc]init];
        datePicker.frame=CGRectMake(0,0,320,300);
        datePicker.tag=100;
        [cell.contentView addSubview:datePicker];
        
        [datePicker addTarget:self action:@selector(dateChanged:) forControlEvents:UIControlEventValueChanged];
    }
    return cell;
}
//将要选中时进行处理
-(NSIndexPath *)tableView:(UITableView *)tableView willSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    //设置只有 due Date 哪一行可以点击
    if(indexPath.section==0&&indexPath.row==2){
        return indexPath;
    }else{
        return nil;
    }
}
//点击某一行进行处理
-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
    [self.textField resignFirstResponder];
    if (indexPath.section==0&&indexPath.row==2) {
        if (!_datePickerVisible) {
            [self showDatePick];
        }else{
            [self hideDatePicker];
        }
    }
}

-(void)cancel{
    [self.delegate itemDetailViewControllerDidCancel:self];
}
-(void)done{
    if (self.itemToEdit.context ==nil) {
//        ChecklistItem *item=[[ChecklistItem alloc]init];
        self.itemToEdit.context=self.textField.text;
        self.itemToEdit.checked=NO;
        self.itemToEdit.shouldRemind=self.switchControl.on;
        self.itemToEdit.dueDate=_dueDate;
//        [self.delegate itemDetailViewController:self didFinishAddingItem:item];
        [_checklistItemService addChecklistItem:self.itemToEdit];
        [self.delegate itemDetailViewControllerDidCancel:self];
    }else{
        self.itemToEdit.context = self.textField.text;
        self.itemToEdit.shouldRemind = self.switchControl.on;
        self.itemToEdit.dueDate = _dueDate;
//        [self.delegate itemDetailViewController:self didFinishEditingItem:self.itemToEdit];
        [_checklistItemService changeChecklistItem:self.itemToEdit];
        [self.delegate itemDetailViewControllerDidCancel:self];
    }
}

//显示选择的日期内容
-(void)updateDueDateLabel{
    NSDateFormatter *formatter=[[NSDateFormatter alloc]init];
    [formatter setDateStyle:NSDateFormatterMediumStyle];
    [formatter setTimeStyle:NSDateFormatterShortStyle];
    self.dueDateLabel.text=[formatter stringFromDate:_dueDate];
}
//获取到选中的日期，并填写到表格中去
-(void)dateChanged:(UIDatePicker*)datePicker{
    _dueDate = datePicker.date;
    [self updateDueDateLabel];
}

//显示日期并进行选择
-(void)showDatePick{
    _datePickerVisible=YES;
    //日期选择一行
    NSIndexPath *indexPathDateRow=[NSIndexPath indexPathForRow:2 inSection:0];
    //新建一行展示日期选项
    NSIndexPath *indexPathDatePicker=[NSIndexPath indexPathForRow:3 inSection:0];
//    UITableViewCell *cell=[self.tableView cellForRowAtIndexPath:indexPathDateRow];
    self.dueDateLabel.tintColor=[UIColor blueColor];
    [self.tableView beginUpdates];
    [self.tableView insertRowsAtIndexPaths:@[indexPathDatePicker]  withRowAnimation:YES];
    [self.tableView reloadRowsAtIndexPaths:@[indexPathDateRow] withRowAnimation:UITableViewRowAnimationNone];
    [self.tableView endUpdates];
    
    UITableViewCell *datePickerCell=[self.tableView cellForRowAtIndexPath:indexPathDatePicker];
    UIDatePicker *datePicker=(UIDatePicker *)[datePickerCell viewWithTag:100];
    [datePicker setDate:_dueDate animated:YES];
}
-(void)hideDatePicker{
    if (_datePickerVisible) {
        _datePickerVisible=NO;
        //日期选择一行
        NSIndexPath *indexPathDateRow=[NSIndexPath indexPathForRow:2 inSection:0];
        //新建一行展示日期选项
        NSIndexPath *indexPathDatePicker=[NSIndexPath indexPathForRow:3 inSection:0];
        //    UITableViewCell *cell=[self.tableView cellForRowAtIndexPath:indexPathDateRow];
        self.dueDateLabel.textColor=[UIColor colorWithWhite:0.0f alpha:0.5f];
        [self.tableView beginUpdates];
        [self.tableView reloadRowsAtIndexPaths:@[indexPathDateRow] withRowAnimation:UITableViewRowAnimationNone];
        [self.tableView deleteRowsAtIndexPaths:@[indexPathDatePicker] withRowAnimation:UITableViewRowAnimationFade];
        [self.tableView endUpdates];
    }
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: ChecklistSwift/ChecklistSwift/ItemDetailViewController.swift
//
//  ItemDetailViewController.swift
//  ChecklistSwift
//
//  Created by 王享 on 15/12/28.
//  Copyright © 2015年 王享. All rights reserved.
//
//获取屏幕宽度
import UIKit

protocol ItemDetailViewControllerDelegate{
    
    func itemDetailViewControllerDidCancel(_ controller:ItemDetailViewController)
    
    func itemDetailViewController(_ controller:ItemDetailViewController,didFinishAddingItem item:ChecklistItem)
    
    func itemDetailViewController(_ controller:ItemDetailViewController, didFinishEditingItem item:ChecklistItem)
}
class ItemDetailViewController: UITableViewController {
    
    var doneBarButton:UIBarButtonItem?
    var textField:UITextField?
    var switchControl:UISwitch?
    var dueDateLabel:UILabel?
    var itemToEdit:ChecklistItem?
    var delegate:ItemDetailViewControllerDelegate?
    
    var _text:String?
    var _shouldRemind:Bool?
    var _datePickerVisible:Bool?
    var _dueDate:Date?
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.textField?.becomeFirstResponder()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.isExclusiveTouch=true
        _datePickerVisible=false
        let cancelBarButton:UIBarButtonItem = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.cancel, target: self, action: #selector(ItemDetailViewController.cancel))
        self.navigationItem.leftBarButtonItem=cancelBarButton;
        
         self.doneBarButton=UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.done, target: self, action: #selector(ItemDetailViewController.done))
         self.navigationItem.rightBarButtonItem=self.doneBarButton;
        
        if (self.itemToEdit != nil) {
            self.title="Edit Item"
            _text=self.itemToEdit!.text
            _shouldRemind=self.itemToEdit!.shouldRemind
            _dueDate=self.itemToEdit!.dueDate as Date?
        }else{
            self.title="Add Item"
            _shouldRemind=false;
            _dueDate=Date.init()
        }
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if( section == 0 && _datePickerVisible!){
            return 4
        }else{
            return 3
        }
    }
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let KScreenWidth = UIScreen.main.bounds.size.width;
        let identifier:String = "Cell"
        var cell:UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: identifier)
        if(cell == nil){
            cell = UITableViewCell.init(style: UITableViewCellStyle.default, reuseIdentifier: identifier)
        }
        if(indexPath.row == 0){
            self.textField = UITextField.init(frame: CGRect(x: 10, y: 7, width: 300, height: 30))
            self.textField?.placeholder="Checklist of Name"
            self.textField?.text=_text
            cell?.contentView.addSubview(self.textField!)
        }else if (indexPath.row==1){
            let label:UILabel = UILabel.init(frame: CGRect(x: 10, y: 7, width: 100, height: 30))
            label.text="Remind Me"
            label.textAlignment=NSTextAlignment.center
            cell?.addSubview(label)
            
            self.switchControl = UISwitch.init(frame: CGRect(x: 250, y: 6, width: 51, height: 31))
            self.switchControl?.isOn=_shouldRemind!
            cell?.addSubview(self.switchControl!)
        }else if(indexPath.row == 2){
            let label:UILabel = UILabel.init(frame: CGRect(x: 10, y: 7, width: 80, height: 30))
            label.text="Due Date"
            label.textAlignment = NSTextAlignment.center
            cell?.contentView.addSubview(label)
            
            self.dueDateLabel = UILabel.init(frame: CGRect(x: 100, y: 7, width: 220, height: 30))
            self.dueDateLabel?.textAlignment=NSTextAlignment.center
            cell?.contentView.addSubview(self.dueDateLabel!)
            self.updateDueDateLabel()
        }else{
            cell?.selectionStyle=UITableViewCellSelectionStyle.none
            let datePicker = UIDatePicker.init(frame: CGRect(x: 0,y: 0,width: KScreenWidth,height: 216))
            datePicker.tag = 100
            cell?.contentView.addSubview(datePicker)
            datePicker.addTarget(self, action: #selector(ItemDetailViewController.dateChanged(_:)), for: UIControlEvents.valueChanged)
        }
        return cell!
    }
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return UITableViewAutomaticDimension
    }
    
    override func tableView(_ tableView: UITableView, willSelectRowAt indexPath: IndexPath) -> IndexPath? {
        if(indexPath.section==0&&indexPath.row==2){
            return indexPath;
        }else{
            return nil;
        }
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        self.tableView.deselectRow(at: indexPath, animated: true)
        self.textField?.resignFirstResponder()
        if (indexPath.section==0&&indexPath.row==2) {
            if (!_datePickerVisible!) {
                self.showDatePick()
            }else{
                self.hideDatePicker()
            }
        }
    }
    func cancel(){
        self.delegate?.itemDetailViewControllerDidCancel(self)
    }
    func done(){
        if (self.itemToEdit==nil) {
            self.itemToEdit = ChecklistItem.init()
            itemToEdit?.text=self.textField?.text
            itemToEdit?.checked=false
            self.itemToEdit?.shouldRemind=self.switchControl?.isOn
            itemToEdit?.dueDate=_dueDate
            self.delegate?.itemDetailViewController(self, didFinishAddingItem: self.itemToEdit!)
        }else{
            self.itemToEdit?.text=self.textField?.text
            self.itemToEdit?.shouldRemind=self.switchControl?.isOn
            self.itemToEdit?.dueDate=_dueDate
            self.delegate?.itemDetailViewController(self, didFinishEditingItem: self.itemToEdit!)
        }
    }
    /**
     显示选择的日期内容
     */
    func updateDueDateLabel(){
        let formatter:DateFormatter = DateFormatter.init()
        formatter.dateStyle=DateFormatter.Style.medium
        formatter.timeStyle=DateFormatter.Style.short
        self.dueDateLabel?.text=formatter.string(from: _dueDate!)
    }
    func dateChanged(_ datePicker:UIDatePicker){
        _dueDate = datePicker.date
        self.updateDueDateLabel()
    }
    
    func showDatePick(){
        _datePickerVisible=true
        let indexPathDateRow:IndexPath = IndexPath(row: 2, section: 0)
        let indexPathDatePicker:IndexPath = IndexPath(row: 3, section: 0)
        self.dueDateLabel?.tintColor=UIColor.blue
        self.tableView.beginUpdates()
        self.tableView.insertRows(at: [indexPathDatePicker], with: UITableViewRowAnimation.automatic)
        self.tableView.reloadRows(at: [indexPathDateRow], with: UITableViewRowAnimation.automatic)
        self.tableView.endUpdates()
        
        let datePickerCell:UITableViewCell? = self.tableView.cellForRow(at: indexPathDatePicker)
        let datePicker:UIDatePicker = datePickerCell!.viewWithTag(100) as! UIDatePicker
        datePicker.setDate(_dueDate!, animated: true)
    }
    
    func hideDatePicker(){
        if(_datePickerVisible!){
            _datePickerVisible = false
            let indexPathDateRow:IndexPath = IndexPath(row: 2, section: 0)
            let indexPathDatePicker:IndexPath = IndexPath(row: 3, section: 0)
            self.dueDateLabel!.textColor=UIColor.darkGray
            self.tableView.beginUpdates()
            self.tableView.deleteRows(at: [indexPathDatePicker], with: UITableViewRowAnimation.automatic)
            self.tableView.reloadRows(at: [indexPathDateRow], with: UITableViewRowAnimation.automatic)
            self.tableView.endUpdates()
        }
    }


    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
}


